[SECURITY PROBLEM] Implement a configuration option to disable the suggestion feature when a GraphQL query fails

tomoikey opened the pull request:
# Purpose
`gqlparser` is highly convenient because it offers suggestions when we mistakenly input field names or type names. However, when deploying a GraphQL server that uses `gqlparser`, leaving the suggestion feature enabled may pose <ins>**a risk of disclosing information to attackers**</ins>. Therefore, it is important to provide developers implementing GraphQL servers with the option to enable or disable the suggestion feature.

For your reference, Rust-based GraphQL library also offer the capability to disable suggestions.
https://github.com/async-graphql/async-graphql/blob/3046ae7f06ac08a9b912d8655d17f5c7f8c663c0/src/schema.rs#L217-L221

I would be delighted if you are pleased with this change. Thank you.

# How to use
### If you want to disable suggestion feature
just add `validator.DisableSuggestion{}`.
```go
gqlparser.LoadQuery(schema, str, validator.DisableSuggestion{})

gqlparser.MustLoadQuery(schema, str, validator.DisableSuggestion{})

validator.Validate(schema, doc, validator.DisableSuggestion{})
```

### If you want to use them the same as previous versions.
there is no changes.
```go
gqlparser.LoadQuery(schema, str)

gqlparser.MustLoadQuery(schema, str)

validator.Validate(schema, doc)
```

---

I have:
 - [x] Added tests covering the bug / feature 
 ~- [ ] Updated any relevant documentation~

> Updated any relevant documentation

Additionally, it was mentioned that updating the relevant documentation is mandatory, but I was unsure about the specific sections that need to be revised. If you find this PR satisfactory, I would greatly appreciate it if you could advise me on which documents to update and how to modify them.

tomoikey commented:
The golangci-lint CI is currently failing, but since the same issue is occurring on the main branch, I will not be addressing this in my branch.

kgrigorev commented:
Hi @tomoikey we have run into similar problem by using this library (that suggesting types in the error message is a security risk for a public GraphQL server). So we are also interested in this change. But I am afraid your change may break a lot of existing code, because of adding additional parameter to the exported functions (which are used by other packages). Have you maybe considered ways for keeping an old API intact, i.e. that gqlparser.MustLoadQuery(schema, query) etc would still work? 

Another idea would be just to drop suggested types from the error message

tomoikey commented:
@kgrigorev 
Nice to meet you! I’m really happy that you’re interested in my changes.
I will respond to your comments below, but since I am not very confident in my English, there may be mistakes or unintended impolite expressions. If anything is unclear, please feel free to ask for clarification.

> gqlparser.MustLoadQuery(schema, query) etc would still work?

In conclusion, you are correct that this change breaks backward compatibility and will cause compilation errors. However, I have implemented it so that <ins>**passing `nil` or the zero value of `validator.ValidateOption` will result in the same behavior as before**</ins>. In fact, <ins>**all the existing tests for gqlparser are passing, with `nil` used as the configuration value in each case**</ins>.

The reasoning and background that led to this implementation are explained in the paragraphs below.

> I am afraid your change may break a lot of existing code, because of adding an additional parameter to the exported functions.

> Have you maybe considered ways for keeping an old API intact

You are absolutely right. When I was considering how to implement a solution to this issue, the biggest concern was maintaining **backward compatibility**. First, let me share the general approach I took with my implementation. Since the suggestion feature is quite useful during development, I aimed not to remove it but rather provide users of the library with a way to choose whether to use it. Here is the approach I had in mind:

1. Maintain backward compatibility by introducing new functions without changing the signatures of the existing public functions.
2. Modify the signatures of the existing public functions, sacrificing backward compatibility but providing a method for users to achieve the same behavior as previous versions with minimal changes.

In more concrete terms, the implementation would look something like this:

1. Introduce new functions such as `LoadQueryWithOption` (or `LoadQueryWithoutSuggestion`), `MustLoadQueryWithOption` (or `MustLoadQueryWithoutSuggestion`), and `ValidateWithOption` (or `ValidateWithoutOption`), while keeping `LoadQuery`, `MustLoadQuery`, and `Validate` unchanged.
2. Add arguments to the signatures of `LoadQuery`, `MustLoadQuery`, and Validate for injecting configurations, ensuring that users can get the same behavior as before simply by passing in empty configuration values.

The pros and cons of each approach are as follows:

* Approach 1:
  * Pros:
    * Maintains backward compatibility, so users can continue using the existing functions without changes, while new functionality can be accessed via new functions.
  * Cons:
    * Increases the number of functions in the library just to maintain backward compatibility.
    * Introduces potential duplication between the old and new functions.
    * There may be a need to deprecate the old functions, which would require additional management and maintenance over time.
* Approach 2:
  * Pros:
    * By simply modifying the signatures of the existing functions, the implementation remains clean and simple.
    * Users can achieve the same behavior as before by passing empty configuration values, so the required changes on their side are minimal.
    * There is no need to worry about future deprecations, making maintenance of the library easier in the long 
run.
  * Cons:
    * Lacks backward compatibility, so users will need to update their code when upgrading the library.
    * We need to inform users in the release notes that they only need to pass nil or the zero value to retain the same behavior as before.

Approach 1 has the significant advantage of maintaining backward compatibility. However, the increased redundancy in the library’s code, and the **potential** need to manage deprecated functions, are notable downsides.

On the other hand, Approach 2 does not preserve backward compatibility, meaning that users will need to update their code. However, considering the typical use cases for the gqlparser library, where it is likely used in specific locations rather than across various parts of the application code, I believe this wouldn’t impose too much of a burden on users. Furthermore, the required change is simply adding an empty configuration value, which makes the downside of Approach 2 relatively minor.
Additionally, with Approach 2, it will allow for flexible modifications in the future if we need to further break down or add configuration options.

For these reasons, I chose to proceed with Approach 2.

If this PR is merged, Users will need to make the following changes to maintain the same behavior as in previous versions.
```diff
- validator.Validate(a, b)
+ validator.Validate(a, b, nil)

- gqlparser.LoadQuery(a, b)
+ gqlparser.LoadQuery(a, b, nil)

- gqlparser.MustLoadQuery(a, b)
+ gqlparser.MustLoadQuery(a, b, nil)
```

> Another idea would be just to drop suggested types from the error message

Thank you for suggesting a new approach! This might overlap with what I mentioned earlier, but I believe the Suggestion feature is very useful, especially in development builds. I feel that providing more detailed error information can help smooth the development process, particularly for those who may not be familiar with the internal workings of the GraphQL server. For this reason, I am concerned that simply removing this feature might make things less convenient.

tomoikey commented:
@kgrigorev
I have come up with a better solution than the one mentioned above. The idea is to use variable-length arguments for each function. This way, even if we update from the previous version, no compilation errors will occur. Furthermore, if nothing is passed to the variable-length arguments, we can implement it to behave the same as before, which will resolve our issue.

```go
func LoadQuery(schema *ast.Schema, str string, options ...validator.ValidateOptionFactor) (*ast.QueryDocument, gqlerror.List)
func MustLoadQuery(schema *ast.Schema, str string, options ...validator.ValidateOptionFactor) *ast.QueryDocument
func Validate(schema *Schema, doc *QueryDocument, options ...ValidateOptionFactor) gqlerror.List
```

fixed on: c01da5c

StevenACoffman commented on validator/imported_test.go:
```diff
@@ -2,6 +2,7 @@ package validator_test
 
 import (
 	"fmt"
+	"github.com/vektah/gqlparser/v2/validator"
```
Would you mind putting this import down with the non-standard library imports?

tomoikey replied:
@StevenACoffman 
Thanks! I fixed it like
```go
import (
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/vektah/gqlparser/v2"
	"github.com/vektah/gqlparser/v2/ast"
	"github.com/vektah/gqlparser/v2/parser"
	"github.com/vektah/gqlparser/v2/validator"
	"github.com/vektah/gqlparser/v2/validator/rules"
)
```

StevenACoffman commented on validator/validator.go:
```diff
@@ -8,7 +8,7 @@ import (
 type AddErrFunc func(options ...ErrorOption)
 
-type ruleFunc func(observers *Events, addError AddErrFunc)
+type ruleFunc func(observers *Events, validateOption ValidateOption, addError AddErrFunc)
```
```suggestion
type ruleFunc func(observers *Events, addError AddErrFunc, options ...ValidateOption)
```
Currently people are using the AddRule which takes a function as an argument that uses this `ruleFunc` signature, so your change here would break backwards compatibility unless you made it variadic.

StevenACoffman commented:
Hey, thanks for working on this. 
There is some tension between avoiding breaking backwards compatibility and adding more configurable behavior. I don't think this is entirely free from breaking changes.

Some recent changes ( #320 ) allow people to reset the rules, so that might be an alternative method that would avoid breaking backwards compatibility.

tomoikey commented:
@StevenACoffman 
Thank you for your review! I’ve incorporated the changes from #320 and adjusted the implementation to introduce a new rule while maintaining full backward compatibility. What are your thoughts on this updated approach?

StevenACoffman commented:
Thanks! Yeah, that's a better way since it avoids breaking backward compatibility.
